Use pytest remotedata

CameronTStark opened the pull request:
<!-- Thank you for your contribution! The following items must be addressed before the code can be merged. Please don't hesitate to ask for help if you're unsure of how to accomplish any of the items. Feel free to remove checklist items that are not relevant to your change. -->

 - [x] Closes #882
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

<!-- Brief description of the problem and proposed solution (if not already fully described in the issue linked to above): -->
This implements pytests-remotedata for tests in pvlib/tests/iotools that require network interactions. This package skips tests marked with `@pytest.mark.remotedata` unless a `--remote-data` flag is included in the `pytest` command. This flag is only implemented in Azure Pipelines for  only the "Test_bare_linux" suite of tests.

In my local tests this has reduced the run time of `pytest pvlib/tests/iotools` from about a minute down to 2-3 seconds. This should result in an overall test time reduction in the CI of 9 minutes. This improvement should contribute significantly towards fixing #878 as well.

kanderso-nrel commented:
Codecov has a point -- is the coverage reported in the front-page badges going to drop because of this?  Can codecov/coveralls be configured to consider those functions covered?

CameronTStark commented:
>Is the coverage reported in the front-page badges going to drop because of this? 

If left alone I believe this will impact this number but if we can point it to coverage results that have used the `--remote-data` flag then it should not. 

> Can codecov/coveralls be configured to consider those functions covered? 

I'm looking into this now but could use some help from someone with experience with codecov.

CameronTStark commented:
The codecov/coveralls pass now. The key was to learn that they work based on Travis and to amend the pytest command with `--remote-data` in the .travis.yml file. Thanks @kanderso-nrel for raising the issue!

wholmgren commented on docs/sphinx/source/whatsnew/v0.7.2.rst:
```diff
@@ -25,11 +25,13 @@ Bug fixes
   passing ``tz=datetime.timezone.utc`` (:pull:`879`)
 * Fix documentation homepage title to "pvlib python" based on first heading on
   the page. (:pull:`890`) (:issue:`888`)
+* Implement pytest-remotedata to increase test suite speed. Requires
```
```suggestion
* Implement `pytest-remotedata <https://github.com/astropy/pytest-remotedata>`_ to increase test suite speed. Requires
```

CameronTStark commented on ci/requirements-py35.yml:
```diff
@@ -6,7 +6,6 @@ dependencies:
     - coveralls
     - cython
     - ephem
-    - netcdf4
```
There was an import error caused by using the conda netcdf4. I'm unsure if it was related to adding pytest-remotedata or if it was a new conda quirk. 

Now that everything is passing I can revert commit 6fa8ecf where this is changed and be doubly sure that it is necessary.

CameronTStark replied:
I reverted and the tests pass so thankfully we can leave netcdf4 where it was.

CameronTStark commented:
> Most of the `test_forecast.py` module could also use the decorator, so maybe we should use the azure `Test_conda_linux` instead of `Test_bare_linux`.

Thanks for pointing out the `test_forecast.py` for adding the decorator. I'm going to work on including that as well. 

I'm sorry I don't quite follow how that affects the choice of `Test_conda_linux` instead of `Test_bare_linux` in Azure. Can you please elaborate?

wholmgren commented:
Those tests need netcdf and siphon so they are skipped in the bare env.

On Thu, Feb 20, 2020 at 4:41 PM Cameron Stark <notifications@github.com>
wrote:

> Most of the test_forecast.py module could also use the decorator, so
> maybe we should use the azure Test_conda_linux instead of Test_bare_linux.
>
> Thanks for pointing out the test_forecast.py for adding the decorator.
> I'm going to work on including that as well.
>
> I'm sorry I don't quite follow how that affects the choice of
> Test_conda_linux instead of Test_bare_linux in Azure. Can you please
> elaborate?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/pvlib/pvlib-python/pull/896?email_source=notifications&email_token=ABBOERZ3HEPMMMXOC7RKWZTRD4IJZA5CNFSM4KYVG2P2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMQ4GHA#issuecomment-589415196>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBOER2UWQ4VKAA2LVZZVRDRD4IJZANCNFSM4KYVG2PQ>
> .
>

CameronTStark commented:
I'm getting errors by running test_forecast.py locally. Is there something I need to do locally, on top of having siphon and netcdf4 installed, to get those to work? 

![Screen Shot 2020-02-21 at 11 15 37 AM](https://user-images.githubusercontent.com/9811038/75060169-8cf8c880-549b-11ea-8a5c-ee68f1fecd5f.png)


[Azure shows warnings coming from those tests.](https://dev.azure.com/solararbiter/pvlib%20python/_build/results?buildId=2773&view=logs&j=06aebe0f-636a-5f93-19d5-0c37b58d55c8&t=0878738c-c430-5911-4cfd-99d3ab677436&l=178) I could use some help making sure this is already working right before trying to optimize it.

wholmgren commented:
are you running into #895?

CameronTStark commented:
> are you running into #895?

Yes, that was it. I saw that issue but didn't connect. Thanks for the quick help!

CameronTStark commented:
I've added `@pytest.mark.remote_data` to tests in test_forecast.py and dropped that test suite time from ~40 seconds to under 1 second. The overall test suite without `--remote-data` runs under 30 seconds now on my machine. 

This PR is ready for review again.

wholmgren commented on ci/requirements-py35-min.yml:
```diff
@@ -8,6 +8,7 @@ dependencies:
     - pytest
     - pytest-cov
     - pytest-mock
+    - pytest-remotedata
```
is the python 3.5 pytest-remotedata package too old? If so delete it here and make the comment below more clear (e.g. "conda package is 0.x, but we need > 0.3.1")

kanderso-nrel commented:
I scrolled through the test suite and didn't see any unmarked network calls.  I haven't used pytest-remotedata before, but I cloned this branch locally and was able to run the test suite with no internet connection, so looks good to me!  My only suggestion is that it might be good to mention the new command line flag in the [Testing docs](https://pvlib-python.readthedocs.io/en/stable/contributing.html#testing). 

Some tests have both `@pandas.util.testing.network` and the new `@remote_data` decorations.  It's not clear to me exactly what the overlap between the two are, even after reading the [docstring](https://github.com/pandas-dev/pandas/blob/master/pandas/_testing.py#L2262).  Is `@network` necessary now that we disable network tests by default?  I'm not saying that a change should be made, I'm just looking to learn :)

CameronTStark commented:
> My only suggestion is that it might be good to mention the new command line flag in the Testing docs.

This is a good idea. I'll draft a blurb in that section. 

> Some tests have both @pandas.util.testing.network and the new @remote_data decorations. It's not clear to me exactly what the overlap between the two are, even after reading the docstring. Is @network necessary now that we disable network tests by default? 

I wondered about the `@network` decorators and looked at it for a second, didn't fully understand it and left them in place. Looking at them again, all it does is check if www.google.com is reachable before running. If it's reachable the test is run and if it isn't then that test is skipped.

I can't think of a reason that it'll hurt anything to leave in. It could be better to reverse the order putting `@network` inside the `@pytest.mark.remote_data` so the network check won't happen unless `--remote-data` is invoked. I could also see it being beneficial the way it is now too. I'll leave it as is unless someone has a compelling case one way or the other.

wholmgren commented:
thanks for the review and testing @kanderso-nrel and thanks for the contribution @CameronTStark.

let's punt on the `@network` decorator for now.
